### PR TITLE
Clip backdrop root (input) image to border box of element

### PR DESCRIPTION
**Pull request by mfreed7**
With this change, the spec is now explicit that the input image for the backdrop root image should be clipped to the border box of the element. Prior to this change, the spec was written such that the
entire backdrop image was to be used as the filter input. For filters that move pixels (e.g. blur), this
meant that neighboring content that is outside the bounds of the element would influence the filtered
content within the element's bounds.

This change is a good one for these reasons:
 - It agrees more closely with existing Webkit/Safari behavior.
 - It improves performance, by filtering fewer pixels in the
   case of pixel-moving filters.
 - Developers have been disappointed with the existing behavior
   of bringing in pixels from outside the bounds of the
   element, and have requested a way to turn this behavior off.
   See the discussion starting at [1].

### CODE REVIEW

**Comment by mstange** [comment]
This seems reasonable, but I think it needs clarification for what to do if the element with the backdrop-filter is transformed. The filter is applied "outside" the transform; but "outside" the transform, the element's border box is not necessarily an axis-aligned rectangle, and it's unclear how to apply a filter to such a shape. Is the clip applied to the axis-aligned bounding box of the shape? Or is the clip applied to the accurate shape, and then the filter is applied to the bounding box of the shape which now contains transparent parts?

Should we also say something about the edge mode of the `blur()` filter? [filter-effects-1](https://drafts.fxtf.org/filter-effects-1/#blurEquivalent) says that the edgeMode used for `blur()` filter depends on where it's used: If used in the `filter` property, the `edgeMode` is `none`, which means it will blur in transparency from the outside. So we could add a sentence that says that, when used in the `backdrop-filter` property, the `blur()` filter uses `edgeMode="duplicate"`, because otherwise the edges would become transparent even with the extra clipping.

(Also, can you add the link for [1] to your comment on this issue?)

**Comment by mfreed7** [comment]
@mstange, thanks for the great comments. I agree that the case of a transform needs to be clarified. I'm open to suggestions on what the best way to go is for this. Note: this was previously somewhat unclear as well, I think. From a user's perspective, it would seem to me that applying the clip to the accurate (non-axis-aligned quad) shape and then applying the filter with transparent outsides makes the most sense. But it's also the most difficult to implement.

Regarding the edge mode comment, I wholeheartedly agree. The backdrop-filter edgeMode should be "duplicate". I will add this to my PR.

And finally, thanks for noticing the missing link:
[1] https://bugs.chromium.org/p/chromium/issues/detail?id=497522#c196

**Comment by mstange** [comment]
Agreed that it was previously unclear as well.
I'm actually not sure what would be the best way to address this. It's not an easy question to answer. Have you checked how Safari / CoreAnimation deals with the transformed case?
My inclination would be "clip to the axis-aligned bounding box", because that should be the easiest to implement. But you're right that this would result in unfortunate interactions in the case where a transformed element has both an outer box-shadow and a backdrop blur.

**Comment by mfreed7** [comment]
@mstange, please take a look. I included both suggestions, and went (for now at least) with the more rigorous definition of transforms. We could still relax it, but I wanted to see how it would read with the full definition.

File: filter-effects-2/Overview.bs [hunk 1]
```diff
@@ -78,6 +78,10 @@ output as its SourceGraphic.
 
 Filter functions must operate in the sRGB color space.
 
+If the filter functions list includes a <a href="https://drafts.fxtf.org/filter-effects-1/#blurEquivalent">blur()</a>
+filter, the blur filter will be applied with edgeMode="duplicate", with the edge defined by the transformed border
```
Comment (chrishtr) [inline filter-effects-2/Overview.bs]:
```suggestion
filter, the blur filter will be applied with edgeMode="duplicate", with the edge defined by the clipped, transformed border
```

File: filter-effects-2/Overview.bs [hunk 2]
```diff
@@ -78,6 +78,10 @@ output as its SourceGraphic.
 
 Filter functions must operate in the sRGB color space.
 
+If the filter functions list includes a <a href="https://drafts.fxtf.org/filter-effects-1/#blurEquivalent">blur()</a>
```
Comment (chrishtr) [inline filter-effects-2/Overview.bs]: Is it just blur() that is pixel-moving?

Comment (AmeliaBR) [reply filter-effects-2/Overview.bs]: For shorthand filters, both `blur()` and `drop-shadow()` move pixels.  The filter list can also include references to SVG filters, which include various ways of extending effects to neighbouring pixels.

But I don't think it's worth making special cases for the other situations.

Comment (mfreed7) [reply filter-effects-2/Overview.bs]: I added a parenthetical "(or any pixel moving filter)" - let me know if that works for you.

**Comment by chrishtr** [comment]
@mstange any more comments before we commit? is the latest text ok by you?

**Comment by mstange** [comment]
I would prefer to revert the addition of " (or other pixel-moving filter)". `blur()` is the only filter for which [the spec](https://drafts.fxtf.org/filter-effects-1/#blurEquivalent) defines an implicit `edgeMode` based on what property the filter function is used in. For the `drop-shadow()` filter function, there is no such note, and I'm not sure if using `edgeMode="duplicate"` for the drop shadow blur would yield desirable results; I think always using the default `edgeMode="none"` for the drop shadow blur will look better. And for SVG filters, most pixel-moving filters have an explicit `edgeMode` attribute which should be respected, and for those that don't have such an attribute it's not defined what it would mean to use `edgeMode="duplicate"`.

Other than that, the text looks good to me. Thanks!

**Comment by mstange** [comment]
Here's a testcase which uses `drop-shadow()` in `backdrop-filter` in order to create a glow effect on the background: https://codepen.io/mstange/pen/dBvJRR

Screen Shot 2019-06-21 at 11 20 01 AM

It's a bit hard to find appropriate uses for this because `drop-shadow` needs some variation in the input's alpha channel in order to do anything interesting, so you need a Backdrop Image which has some transparent parts, which you can only get if you trigger a non-root Backdrop Root, e.g. using opacity.

In this testcase the glow would look a bit better along the edges if the input was *not* clipped to the element's border box, and if the shadow blur could actually sample from a Backdrop Image that extends beyond the border box. But it's hard to notice the difference.

**Comment by mstange** [comment]
Oh, and now that I've actually tried [the testcase](https://codepen.io/fchristant/pen/VOrYYz) from [the bug comment you linked to](https://bugs.chromium.org/p/chromium/issues/detail?id=497522#c196), I'm no longer sure if specifying `edgeMode="duplicate"` for the `blur()` filter function was such a good idea. In the testcase, if you slowly scroll the contents, in Chrome Canary the blur produces a flickering effect near the top and bottom edges of the header. That's because the pixels along the edge now have a much stronger influence on the blur results than before. There is a lot less flickering in Safari.

**Comment by mstange** [comment]
I have a more fundamental question actually: If we don't clip the inputs to the border-box, and apply backdrop-filter to an element which also has box-shadow, is that box-shadow part of the backdrop image? It gets painted "below" the backdrop-filter in painting order.

**Comment by mfreed7** [comment]
@mstange, thanks for the comments. I put the language back to just "blur" for the edgemode comment.

I've never quite been able to wrap my head around a drop-shadow backdrop-filter, for a couple reasons. First, it is offset from the original, and the backdrop-filter effect is clipped at the border box. Second, the offset/blurred drop shadow is already specified to be composited *below* the image. So if you're applying a backdrop-filtered drop-shadow, where does the drop shadow composite, exactly? I actually thought about explicitly removing drop-shadow from the spec for backdrop-filter, but I figured someone would come up with a cool use.

Finally, regarding the clipping of the input rect: I did mention this problem on the crbug.com/497522 discussion. As soon as one pixel of the new color passes under the border rect, it can "suddenly" affect a large portion of the filtered output. There's no perfect answer to this problem - there are weird behaviors either way. It seems like the developer feedback is that they would prefer the new behavior - clipping the input. It also has the added benefit of being more performant and being closer to Webkit. I'm not sure how their implementation (which does appear to clip the input to the border box, or something close) works better in this regard - let me know if you know!

**Pull request merged**